File: PairArray.ts
// A random-access collection of pairs.
class PairArray<First, Second> {
  // The first part of each element.
  private first: First[] = [];

  // The second part of each element.
  private second: Second[] = [];

  // Creates an empty instance.
  public constructor() {}

  // The `i`th element.
  public get(i: number): [First, Second] {
    return [this.first[i], this.second[i]];
  }

  // The length.
  public get length(): number { return this.first.length; }

  // Adds x to the end.
  public push(x: [First, Second]) {
    this.first.push(x[0]);
    this.second.push(x[1]);
  }
}
const x = new PairArray<number, number>();
x.push([1, 2]);
console.log(x.get(0));

File: PairArray.swift
/// A random-access collection of `(First, Second)` pairs.
struct PairArray<First, Second> {
  /// The first part of each element.
  private var first: [First] = []

  /// The second part of each element.
  private var second: [Second] = []

  /// An empty instance.
  public init() {}

  /// The `i`th element.
  public subscript(i: Int) -> (First, Second) {
    (first[i], second[i])
  }

  /// The length.
  public var count: Int { first.count }

  /// Adds x to the end.
  public mutating func append(_ x: (First, Second)) {
    first.append(x.0)
    second.append(x.1)
  }
}
